drivers/flash/flash_mspi_nor_mx.c

```diff
@@ -0,0 +1,989 @@
+	 * in cfg_mode SPI/OPI and cfg_rate transfer STR/DTR
+	 */
+	ret = flash_mspi_nor_mx_status_read(flash, NOR_MX_STATUS_MEM_ERASED);
+
+	return ret;
+}
+
+static int flash_mspi_nor_mx_erase_block(const struct device *flash, off_t addr)
+{
+	int ret;
+
+	LOG_DBG("Erasing block at 0x%08zx", (ssize_t)addr);
+
+	ret = flash_mspi_nor_mx_command_write(flash, SPI_NOR_CMD_BE, addr, 4, 0, NULL, 0);
+
+	return ret;
+}
+
+static int flash_mspi_nor_mx_erase_chip(const struct device *flash)
+{
+	int ret;
+
+	LOG_DBG("Erasing chip");
+
+	ret = flash_mspi_nor_mx_command_write(flash, SPI_NOR_CMD_CE, 0, 4, 0, NULL, 0);
+
+	return ret;
+}
+
+static int flash_mspi_nor_mx_page_program(const struct device *flash, off_t offset, void *wdata,
+					  size_t len)
+{
+	const struct flash_mspi_nor_mx_config *cfg = flash->config;
+	struct flash_mspi_nor_mx_data *data = flash->data;
+	int ret;
+
+	data->packet.dir = MSPI_TX;
+	data->packet.cmd = data->dev_cfg.write_cmd;
+	data->packet.address = offset;
+	data->packet.data_buf = wdata;
+	data->packet.num_bytes = len;
+
+	data->trans.async = true;         /* use callback on Irq if PIO, meaningless with DMA */
+	data->trans.xfer_mode = MSPI_PIO; /* TODO : transfer with DMA */
+	data->trans.tx_dummy = data->dev_cfg.tx_dummy;
+	data->trans.cmd_length = data->dev_cfg.cmd_length;
+	data->trans.addr_length = data->dev_cfg.addr_length;
+	data->trans.hold_ce = false;
+	data->trans.priority = 1;
+	data->trans.packets = &data->packet;
+	data->trans.num_packet = 1;
+	data->trans.timeout = CONFIG_MSPI_COMPLETION_TIMEOUT_TOLERANCE;
+
+	LOG_DBG("Page programming %d bytes to 0x%08zx", len, (ssize_t)offset);
+
+	ret = mspi_transceive(cfg->bus, &cfg->dev_id, (const struct mspi_xfer *)&data->trans);
+	if (ret) {
+		LOG_ERR("MSPI write transaction failed with code: %d", ret);
+		return -EIO;
+	}
+
+	return ret;
+}
+
+/*
+ * This function Polls the WIP(Write In Progress) bit to become to 0
+ * in nor_mode SPI/OPI XSPI_SPI_MODE or XSPI_OCTO_MODE
+ * and nor_rate transfer STR/DTR XSPI_STR_TRANSFER or XSPI_DTR_TRANSFER
+ */
+static int flash_mspi_nor_mx_mem_ready(const struct device *flash)
+{
+	const struct flash_mspi_nor_mx_config *cfg = flash->config;
+	struct flash_mspi_nor_mx_data *data = flash->data;
+	mspi_timing_cfg bkp = data->timing_cfg;
+	int ret;
+
+	LOG_DBG("Reading status register");
+	ret = flash_mspi_nor_mx_status_read(flash, NOR_MX_STATUS_MEM_RDY);
+	if (ret) {
+		LOG_ERR("Could not read status");
+		return ret;
+	}
+
+	if (data->dev_cfg.io_mode != MSPI_IO_MODE_SINGLE) {
+		data->timing_cfg = bkp;
+		if (mspi_timing_config(cfg->bus, &cfg->dev_id, cfg->timing_cfg_mask,
+				       (void *)&data->timing_cfg)) {
+			LOG_ERR("Failed to config mspi controller");
+			return -EIO;
+		}
+	}
+
+	return ret;
+}
+
+/* Function to read the flash with possible PIO IT or DMA */
+static int flash_mspi_nor_mx_read(const struct device *flash, off_t offset, void *rdata,
+				  size_t size)
```

Comment:
You will need to consider controllers that does not support XIP read.
The config CONFIG_MSPI_XIP is used to indicate a controller's support on that.
It was my bad that I did not include those config in my PR for the atxp032 flash driver and will be fixed in the near future. You should consider adding them like this https://github.com/zephyrproject-rtos/zephyr/pull/79513